[llvm-objcopy] Add change-section-lma *+/-offset

eleanor-arm opened the pull request:
llvm-objcopy did not support change-section-lma argument.

This patch adds support for a use case of change-section-lma, that is
shifting load address of all sections by the same offset. This seems to
be the only practical use case of change-section-lma, found in other
software such as Zephyr RTOS's build system.

This is an option that could possibly be supported in some other than
ELF formats, however this change only implements it for ELF. When used
with other formats an error message is raised.

In comparison, the behavior of GNU objcopy is inconsistent. For some ELF
files it behaves the same as described above. For others, it copies the
file without modifying the p_paddr fields when it would be expected. In
some experiments it modifies arbitrary fields in section or program
headers. It is unclear what exactly determines this.
The executable file generated by yaml2obj in this test is not parsable
by GNU objcopy. With Machine set to EM_AARCH64, the file can be parsed
and the first test in the test file completes with 0 exit code. However,
the result is rather arbitrary. AArch64 GNU objcopy subtracts 0x1000
from p_filesz and p_memsz of the first LOAD section and 0x1000 from
p_offset of the second LOAD section. It does not look meaningful.

eleanor-arm committed:
[llvm-objcopy] Add change-section-lma *+/-offset

llvm-objcopy did not support change-section-lma argument.

This patch adds support for a use case of change-section-lma, that is
shifting load address of all sections by the same offset. This seems to
be the only practical use case of change-section-lma, found in other
software.

jh7370 commented on llvm/tools/llvm-objcopy/ObjcopyOptions.cpp:
```diff
@@ -552,6 +552,33 @@ static Error loadNewSectionData(StringRef ArgValue, StringRef OptionName,
+  } else {
+    return createStringError(errc::invalid_argument,
+                             "bad format for " + OptionName +
+                                 ": it is required that all sections "
```
I think you should rephrase this to something like "changing a specific section LMA is not supported. Use *+val or *-val instead"

eleanor-arm replied:
Fixed

jh7370 commented on llvm/tools/llvm-objcopy/ObjcopyOptions.cpp:
```diff
@@ -552,6 +552,33 @@ static Error loadNewSectionData(StringRef ArgValue, StringRef OptionName,
+  if (StringValue.empty())
+    return createStringError(errc::invalid_argument,
+                             "bad format for " + OptionName +
+                                 ": missing offset of LMA");
```
```suggestion
                                 ": missing LMA offset");
```
would read slightly better

eleanor-arm replied:
Fixed

jh7370 commented on llvm/tools/llvm-objcopy/ObjcopyOptions.cpp:
```diff
@@ -552,6 +552,33 @@ static Error loadNewSectionData(StringRef ArgValue, StringRef OptionName,
+  auto SLMAV = getAsInteger<int64_t>(StringValue);
+  if (!SLMAV)
+    return createStringError(SLMAV.getError(),
+                             "Unable to parse adjustment value");
```
Please see the coding standards for how to format error messages.

eleanor-arm replied:
Fixed

jh7370 commented on llvm/tools/llvm-objcopy/ObjcopyOptions.cpp:
```diff
@@ -552,6 +552,33 @@ static Error loadNewSectionData(StringRef ArgValue, StringRef OptionName,
+                             "bad format for " + OptionName +
+                                 ": missing offset of LMA");
+
+  auto SLMAV = getAsInteger<int64_t>(StringValue);
```
I have no idea what `SLMAV` is supposed to mean. Please don't abbreviate variable names unnecessarily.

eleanor-arm replied:
Done

jh7370 commented on llvm/test/tools/llvm-objcopy/ELF/change-section-lma-single-load.test:
```diff
@@ -0,0 +1,45 @@
+# RUN: yaml2obj %s -o %t
```
It would be preferable to have all test cases in a single file. You can use yaml2obj's --docnum option to have multiple different YAML inputs in the same file (see other tests for examples).

That being said, what's the benefit of the "single-load" test case that the "multiple-load" case doesn't provide?

eleanor-arm replied:
One yaml file seems sufficient. The fixup has only one.

jh7370 commented on llvm/test/tools/llvm-objcopy/ELF/change-section-lma-multiple-load.test:
```diff
@@ -0,0 +1,52 @@
+  Data:            ELFDATA2LSB
+  Type:            ET_EXEC
+  Machine:         EM_AARCH64
+Sections:
```
If I'm not mistaken, I think the sections are actually irrelevant to the test case and code behaviour (you can explicitly set the p_filesz of the segment in the YAML, without needing sections to drive it). If so, let's omit them.

eleanor-arm replied:
When I removed the sections the executable generated by yaml2obj was not usable by llvm-objcopy, although it looked quite similar. 

` error: program header with offset 0x1000 and file size 0x2000 goes past the end of the file`

jh7370 replied:
Okay, fair enough. I've actually changed my mind about not having any sections anyway, as I feel like it would be good to verify that the section sh_addr value isn't changed, as part of "documenting what the feature does through tests".

eleanor-arm replied:
Hopefully the latest version of the test shows this.

jh7370 commented on llvm/test/tools/llvm-objcopy/ELF/change-section-lma-multiple-load.test:
```diff
@@ -0,0 +1,52 @@
+  Class:           ELFCLASS64
+  Data:            ELFDATA2LSB
+  Type:            ET_EXEC
+  Machine:         EM_AARCH64
```
The `Machine` value is irrelevant for the test case, so you can omit it.

eleanor-arm replied:
Done

jh7370 commented on llvm/test/tools/llvm-objcopy/ELF/change-section-lma-multiple-load.test:
```diff
@@ -0,0 +1,52 @@
+    Size:            0x1000
+ProgramHeaders:
+  - Type:     PT_LOAD
+    Flags:    [ PF_R ]
```
I believe the `Flags` are irrelevant, so please omit them, to reduce the noise in the test.

eleanor-arm replied:
Done

jh7370 commented on llvm/test/tools/llvm-objcopy/ELF/change-section-lma-multiple-load.test:
```diff
@@ -0,0 +1,52 @@
+    VAddr:    0x1000
+    FirstSec: .text1
+    LastSec:  .text2
+# CHECK-PLUS:  Type           Offset   VirtAddr           PhysAddr           FileSiz  MemSiz
```
Once you've simplified the YAML as suggested, it would be more readable to order the test something like this, I think:
```
# RUN: ...
# RUN: ...

# CHECK...
# CHECK...

<yaml>
```

eleanor-arm replied:
Interestingly, I was told to add the checks inline when contributing to other parts of LLVM but I don't mind them being all on top.

jh7370 commented on llvm/test/tools/llvm-objcopy/ELF/change-section-lma-multiple-load.test:
```diff
@@ -0,0 +1,52 @@
+# RUN: yaml2obj %s -o %t
```
I think you additionally need:
1) A segment with 0 file size.
2) A segment with an explicit p_paddr that is different to the p_vaddr.
3) A non PT_LOAD segment (with non-zero p_filesz), to show that it's not just PT_LOAD segments that are impacted (should they be impacted though? what does GNU objcopy do?).
4) A nested segment (with non-zero p_filesz) to show that segments within other segments are appropriately updated. The segment should probably have a non-zero offset within the PT_LOAD it's nested inside.
5) A segment with 0 p_vaddr/p_paddr.
6) Test cases that cover all the different new error messages you've added for parsing the option.

eleanor-arm replied:
In change-section-lma.test: 
1. PT_NOTE
2. PT_LOAD (.text2) 
3. PT_PHDR
4. PT_PHDR in PR_LOAD (.text1) 
5. PT_NOTE
6. Covers all of them but the use with unsupported file formats which seems excessive

eleanor-arm replied:
> The segment should probably have a non-zero offset within the PT_LOAD it's nested inside.
I am not sure how that helps with coverage.

jh7370 replied:
I honestly can't remember what my reasoning was with that case 🤷‍♂️

jh7370 replied:
With point 3, I was explicitly referring to a segment not nested within another segment. This is important, because some tools treat PT_LOAD segments (and segments nested within them) as different to other top-level segment types, hence why I'd ideally like this checked against GNU's behaviour.

eleanor-arm replied:
I see. For point 3, I will add another non-zero-size NOTE header in fixup 4 that is independent from the LOAD headers.
As mentioned before in other places I don't see a meaning in GNU's objcopy actual behaviour with this particular option. However, I defined what the LLVM version will do based on GNU documentation, some bits and pieces of GNU behaviour and how I saw the option being used in real software. 

The zero-size NOTE is already independent as it automatically gets virtual address of 0x0.

jh7370 replied:
The non-empty PT_NOTE is currently nested inside the first PT_LOAD segment, if I'm not mistaken? I think it needs to have a companion section that can be assigned to it, to force the offset to be not within the other PT_LOADs.

> The zero-size NOTE is already independent as it automatically gets virtual address of 0x0.

Not quite what I meant (I wanted a segment with 0 virtual address, and a different segment with zero file size). That being said, I actually don't think we need a 0 virtual address case, having given it a bit more thought, since there's nothing special about this case.

eleanor-arm replied:
Ah OK, I've changed the test so that the non-zero PT_NOTE section is elsewhere in the file. And then I thought it was better to make the 0-size segment independent too (it might not really matter considering it is zero size).

jh7370 commented on llvm/include/llvm/ObjCopy/CommonConfig.h:
```diff
@@ -245,6 +245,9 @@ struct CommonConfig {
   SmallVector<NewSymbolInfo, 0> SymbolsToAdd;
 
+  // Integer options
+  int64_t ChangeSectionLMAValAll = 0;
```
Should this option really be in `CommonConfig.h` and not the ELF-specific version? Is it actually applicable to non-ELF formats?

eleanor-arm replied:
Possibly applicable but not implemented by this patch. I left it there for now, and added more explanation in the commit message and added the option to ConfigManager.cpp

jh7370 commented on llvm/tools/llvm-objcopy/ObjcopyOpts.td:
```diff
@@ -254,6 +254,14 @@ def adjust_start : JoinedOrSeparate<["--"], "adjust-start">,
 
+defm adjust_section_lma
+    : Eq<"change-section-lma",
+         "Change LMA of all sections by <val>. LMA is the physical address where the section is "
+         "loaded in the memory; as opoosed to VMA, which is the virtual address at which the "
+         "section resides during program execution. Ussually LMA and VMA are the same, but some "
+         "platforms support these to be different resulting in runtime copying">,
```
This description seems too verbose for the help text, and should be in the documentation instead, if needed at all.

eleanor-arm replied:
I've made the description simpler.

jh7370 commented on llvm/tools/llvm-objcopy/ObjcopyOpts.td:
```diff
@@ -254,6 +254,14 @@ def adjust_start : JoinedOrSeparate<["--"], "adjust-start">,
+         "loaded in the memory; as opoosed to VMA, which is the virtual address at which the "
+         "section resides during program execution. Ussually LMA and VMA are the same, but some "
+         "platforms support these to be different resulting in runtime copying">,
+      MetaVarName<"*{+|-}val">;
```
How does this appear in the --help output?

eleanor-arm replied:
It looks as expected.

jh7370 reviewed (commented):
Could you describe what the actual behaviour is when GNU objcopy uses this option, please?

Also, could you go into more detail about what your use-case is, please?

eleanor-arm committed:
fixup1! [llvm-objcopy] Add change-section-lma *+/-offset

llvm-objcopy did not support change-section-lma argument.

This patch adds support for a use case of change-section-lma, that is
shifting load address of all sections by the same offset. This seems to
be the only practical use case of change-section-lma, found in other
software such as Zephyr RTOS's build system.

This is an option that could possibly be supported in some other than
ELF formats, however this change only implements it for ELF. When used
with other formats an error message is raised.

In comparison, the behavior of GNU objcopy is inconsistent. For some ELF
files it behaves the same as described above. For others, it copies the
file without modifying the p_paddr fields when it would be expected. In
some experiments it modifies arbitrary fields in section or program
headers. It is unclear what exactly determines this.
The executable file generated by yaml2obj in this test is not parsable
by GNU objcopy. With Machine set to EM_AARCH64, the file can be parsed
and the first test in the test file completes with 0 exit code. However,
the result is rather arbitrary. AArch64 GNU objcopy subtracts 0x1000
from p_filesz and p_memsz of the first LOAD section and 0x1000 from
p_offset of the second LOAD section. It does not look meaningful.

eleanor-arm commented on llvm/docs/CommandGuide/llvm-objcopy.rst:
```diff
@@ -528,6 +528,10 @@ them.
 
+.. option:: --change-section-lma \*{+-}<offset>
+
+ Currently only supports changing LMA for all sections by the same offset.
```
Done

eleanor-arm commented on llvm/lib/ObjCopy/ELF/ELFObjcopy.cpp:
```diff
@@ -670,6 +670,13 @@ static Error handleArgs(const CommonConfig &Config, const ELFConfig &ELFConfig,
   }
 
+  if (Config.ChangeSectionLMAValAll != 0) {
+    for (auto &Seg : Obj.segments()) {
```
Fixed

eleanor-arm commented on llvm/tools/llvm-objcopy/ObjcopyOptions.cpp:
```diff
@@ -552,6 +552,38 @@ static Error loadNewSectionData(StringRef ArgValue, StringRef OptionName,
+    return createStringError(LMAValue.getError(),
+                             "bad format for " + OptionName + ": value after " +
+                                 ArgValue.slice(0, 2) + " is " + StringValue +
+                                 " when it should be integer");
```
This is intentional. "integer" is meant as an adjective here as the value is obviously a string. It just needs to be convertible to an integer.

jh7370 replied:
"integer" isn't an adjective though, it's a noun. You can use "integral", I guess, but I think the meaning is clear enough and more grammatically correct with my suggestion.

eleanor-arm replied:
OK. I've changed it to the original suggestion.

eleanor-arm committed:
fixup2! [llvm-objcopy] Add change-section-lma *+/-offset

llvm-objcopy did not support change-section-lma argument.

This patch adds support for a use case of change-section-lma, that is
shifting load address of all sections by the same offset. This seems to
be the only practical use case of change-section-lma, found in other
software such as Zephyr RTOS's build system.

This is an option that could possibly be supported in some other than
ELF formats, however this change only implements it for ELF. When used
with other formats an error message is raised.

In comparison, the behavior of GNU objcopy is inconsistent. For some ELF
files it behaves the same as described above. For others, it copies the
file without modifying the p_paddr fields when it would be expected. In
some experiments it modifies arbitrary fields in section or program
headers. It is unclear what exactly determines this.
The executable file generated by yaml2obj in this test is not parsable
by GNU objcopy. With Machine set to EM_AARCH64, the file can be parsed
and the first test in the test file completes with 0 exit code. However,
the result is rather arbitrary. AArch64 GNU objcopy subtracts 0x1000
from p_filesz and p_memsz of the first LOAD section and 0x1000 from
p_offset of the second LOAD section. It does not look meaningful.

eleanor-arm commented on llvm/docs/CommandGuide/llvm-objcopy.rst:
```diff
@@ -528,6 +528,10 @@ them.
 
+.. option:: --change-section-lma \*{+-}<val>
+
+ Shift LMA of non-zero-sized sections in the program header by ``<val>``
```
Fixed

eleanor-arm commented on llvm/docs/CommandGuide/llvm-objcopy.rst:
```diff
@@ -528,6 +528,10 @@ them.
 
  Mark all defined global symbols as weak in the output.
 
+.. option:: --change-section-lma \*{+-}<val>
```
Fixed

eleanor-arm commented on llvm/test/tools/llvm-objcopy/ELF/change-section-lma.test:
```diff
@@ -0,0 +1,55 @@
+# RUN: yaml2obj %s -o %t
+# RUN: llvm-objcopy --change-section-lma *+0x20 %t %t2
+# RUN: llvm-readelf -l %t2 | FileCheck %s --check-prefix=CHECK-PLUS
+# RUN: llvm-objcopy --change-section-lma *-0x10 %t2 %t3
```
Changed, thanks for the suggestion

eleanor-arm commented on llvm/lib/ObjCopy/ELF/ELFObjcopy.cpp:
```diff
@@ -670,6 +670,13 @@ static Error handleArgs(const CommonConfig &Config, const ELFConfig &ELFConfig,
+  if (Config.ChangeSectionLMAValAll != 0) {
+    for (Segment &Seg : Obj.segments()) {
+      if (Seg.FileSize > 0)
+        Seg.PAddr += Config.ChangeSectionLMAValAll;
```
That's a very good point. I've added code to check for overflow/underflow.

jh7370 commented on llvm/lib/ObjCopy/ELF/ELFObjcopy.cpp:
```diff
@@ -670,6 +670,27 @@ static Error handleArgs(const CommonConfig &Config, const ELFConfig &ELFConfig,
+            Seg.PAddr > std::numeric_limits<uint64_t>::max() -
+                            Config.ChangeSectionLMAValAll) {
+          return createStringError(errc::invalid_argument,
+                                   "address 0x" + Twine::utohexstr(Seg.PAddr) +
```
For this and the similar error below, I think it would be useful to give a little more context to the error, something like "would overflow when adjusting by <value>".

eleanor-arm replied:
Improved this as well.

jh7370 commented on llvm/lib/ObjCopy/ELF/ELFObjcopy.cpp:
```diff
@@ -670,6 +670,27 @@ static Error handleArgs(const CommonConfig &Config, const ELFConfig &ELFConfig,
+          return createStringError(errc::invalid_argument,
+                                   "address 0x" + Twine::utohexstr(Seg.PAddr) +
+                                       " would underflow");
+        else
```
I believe LLVM coding standards direct that this `else` should have braces, since the other parts of the if/else chain do.

eleanor-arm replied:
The coding standard actually discourages use of `else` after return, so I've removed it. And I've added braces around the previous `else if` because the control flow is "deeply nested".

jh7370 commented on llvm/docs/CommandGuide/llvm-objcopy.rst:
```diff
@@ -299,6 +299,10 @@ them.
 
+.. option:: --change-section-lma \*{+-}<val>
+
+ Shift LMA of non-zero-sized sections in the program header table by ``<val>``
```
```suggestion
 Shift LMA of non-zero-sized segments by ``<val>``
```
How about this? Sections aren't in the program header table, only program headers (aka segments, which are in turn implicitly in the program header table).

eleanor-arm replied:
Added this suggestion to **Fixup3**

eleanor-arm committed:
fixup3! [llvm-objcopy] Add change-section-lma *+/-offset

llvm-objcopy did not support change-section-lma argument.

This patch adds support for a use case of change-section-lma, that is
shifting load address of all sections by the same offset. This seems to
be the only practical use case of change-section-lma, found in other
software such as Zephyr RTOS's build system.

This is an option that could possibly be supported in some other than
ELF formats, however this change only implements it for ELF. When used
with other formats an error message is raised.

In comparison, the behavior of GNU objcopy is inconsistent. For some ELF
files it behaves the same as described above. For others, it copies the
file without modifying the p_paddr fields when it would be expected. In
some experiments it modifies arbitrary fields in section or program
headers. It is unclear what exactly determines this.
The executable file generated by yaml2obj in this test is not parsable
by GNU objcopy. With Machine set to EM_AARCH64, the file can be parsed
and the first test in the test file completes with 0 exit code. However,
the result is rather arbitrary. AArch64 GNU objcopy subtracts 0x1000
from p_filesz and p_memsz of the first LOAD section and 0x1000 from
p_offset of the second LOAD section. It does not look meaningful.

eleanor-arm commented:
>  it would be good to verify that the section sh_addr value isn't changed, as part of "documenting what the feature does through tests".

I've tried to add this to the test. Is it clear enough what it's showing?

jh7370 commented on llvm/test/tools/llvm-objcopy/ELF/change-section-lma.test:
```diff
@@ -1,32 +1,45 @@
 # RUN: yaml2obj %s -o %t
 # RUN: llvm-objcopy --change-section-lma *+0x20 %t %t2
-# RUN: llvm-readelf -l %t2 | FileCheck %s --check-prefix=CHECK-PLUS
+# RUN: llvm-readelf --program-headers %t2 | FileCheck %s --check-prefix=CHECK-PLUS-PROGRAMS
+# RUN: llvm-readelf --section-headers %t2 | FileCheck %s --check-prefix=CHECK-PLUS-SECTIONS
```
You could simplify this and the similar below down to a single call with both options specified.

eleanor-arm replied:
Yes the original version had them in one command but I "simplified" it by having two separate commands. I think it makes it clearly to see what is being checked.

jh7370 reviewed (commented):
This is mostly fine, but it looks like my last round of test comments haven't been addressed yet?

eleanor-arm committed:
fixup4! [llvm-objcopy] Add change-section-lma *+/-offset

llvm-objcopy did not support change-section-lma argument.

This patch adds support for a use case of change-section-lma, that is
shifting load address of all sections by the same offset. This seems to
be the only practical use case of change-section-lma, found in other
software such as Zephyr RTOS's build system.

This is an option that could possibly be supported in some other than
ELF formats, however this change only implements it for ELF. When used
with other formats an error message is raised.

In comparison, the behavior of GNU objcopy is inconsistent. For some ELF
files it behaves the same as described above. For others, it copies the
file without modifying the p_paddr fields when it would be expected. In
some experiments it modifies arbitrary fields in section or program
headers. It is unclear what exactly determines this.
The executable file generated by yaml2obj in this test is not parsable
by GNU objcopy. With Machine set to EM_AARCH64, the file can be parsed
and the first test in the test file completes with 0 exit code. However,
the result is rather arbitrary. AArch64 GNU objcopy subtracts 0x1000
from p_filesz and p_memsz of the first LOAD section and 0x1000 from
p_offset of the second LOAD section. It does not look meaningful.

eleanor-arm commented:
I've responded to all the unresolved comments. Hopefully, this should cover it all.

eleanor-arm committed:
fixup5! [llvm-objcopy] Add change-section-lma *+/-offset

llvm-objcopy did not support change-section-lma argument.

This patch adds support for a use case of change-section-lma, that is
shifting load address of all sections by the same offset. This seems to
be the only practical use case of change-section-lma, found in other
software such as Zephyr RTOS's build system.

This is an option that could possibly be supported in some other than
ELF formats, however this change only implements it for ELF. When used
with other formats an error message is raised.

In comparison, the behavior of GNU objcopy is inconsistent. For some ELF
files it behaves the same as described above. For others, it copies the
file without modifying the p_paddr fields when it would be expected. In
some experiments it modifies arbitrary fields in section or program
headers. It is unclear what exactly determines this.
The executable file generated by yaml2obj in this test is not parsable
by GNU objcopy. With Machine set to EM_AARCH64, the file can be parsed
and the first test in the test file completes with 0 exit code. However,
the result is rather arbitrary. AArch64 GNU objcopy subtracts 0x1000
from p_filesz and p_memsz of the first LOAD section and 0x1000 from
p_offset of the second LOAD section. It does not look meaningful.

eleanor-arm committed:
fixup6! [llvm-objcopy] Add change-section-lma *+/-offset

llvm-objcopy did not support change-section-lma argument.

This patch adds support for a use case of change-section-lma, that is
shifting load address of all sections by the same offset. This seems to
be the only practical use case of change-section-lma, found in other
software such as Zephyr RTOS's build system.

This is an option that could possibly be supported in some other than
ELF formats, however this change only implements it for ELF. When used
with other formats an error message is raised.

In comparison, the behavior of GNU objcopy is inconsistent. For some ELF
files it behaves the same as described above. For others, it copies the
file without modifying the p_paddr fields when it would be expected. In
some experiments it modifies arbitrary fields in section or program
headers. It is unclear what exactly determines this.
The executable file generated by yaml2obj in this test is not parsable
by GNU objcopy. With Machine set to EM_AARCH64, the file can be parsed
and the first test in the test file completes with 0 exit code. However,
the result is rather arbitrary. AArch64 GNU objcopy subtracts 0x1000
from p_filesz and p_memsz of the first LOAD section and 0x1000 from
p_offset of the second LOAD section. It does not look meaningful.

jh7370 commented on llvm/docs/CommandGuide/llvm-objcopy.rst:
```diff
@@ -299,6 +299,10 @@ them.
 
+.. option:: --change-section-lma \*{+-}<val>
+
+ Shift LMA of non-zero-sized segments by ``<val>``
```
```suggestion
 Shift LMA of non-zero-sized segments by ``<val>``.
```

jh7370 reviewed (approved):
One nit, otherwise LGTM.

eleanor-arm committed:
fixup7! [llvm-objcopy] Add change-section-lma *+/-offset

llvm-objcopy did not support change-section-lma argument.

This patch adds support for a use case of change-section-lma, that is
shifting load address of all sections by the same offset. This seems to
be the only practical use case of change-section-lma, found in other
software such as Zephyr RTOS's build system.

This is an option that could possibly be supported in some other than
ELF formats, however this change only implements it for ELF. When used
with other formats an error message is raised.

In comparison, the behavior of GNU objcopy is inconsistent. For some ELF
files it behaves the same as described above. For others, it copies the
file without modifying the p_paddr fields when it would be expected. In
some experiments it modifies arbitrary fields in section or program
headers. It is unclear what exactly determines this.
The executable file generated by yaml2obj in this test is not parsable
by GNU objcopy. With Machine set to EM_AARCH64, the file can be parsed
and the first test in the test file completes with 0 exit code. However,
the result is rather arbitrary. AArch64 GNU objcopy subtracts 0x1000
from p_filesz and p_memsz of the first LOAD section and 0x1000 from
p_offset of the second LOAD section. It does not look meaningful.
